File: src/main/java/com/github/mgramin/sqlboot/resource_type/impl/sql/SqlResourceType.java
package com.github.mgramin.sqlboot.resource_type.impl.sql;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import com.github.mgramin.sqlboot.exceptions.BootException;
import com.github.mgramin.sqlboot.model.DbResource;
import com.github.mgramin.sqlboot.model.DbResourceThin;
import com.github.mgramin.sqlboot.resource_type.ResourceType;
import com.github.mgramin.sqlboot.sql.ISqlHelper;
import com.github.mgramin.sqlboot.uri.Uri;
import com.github.mgramin.sqlboot.uri.impl.DbUri;
import org.apache.log4j.Logger;
import static java.util.Optional.ofNullable;
import static org.apache.commons.lang3.StringUtils.strip;

/**
 * Created by MGramin on 12.07.2017.
 */
public class SqlResourceType implements ResourceType {

    private final static Logger logger = Logger.getLogger(SqlResourceType.class);

    private final ISqlHelper sqlHelper;
    private final List<String> aliases;
    private final List<ResourceType> child;
    private final String sql;

    public SqlResourceType(ISqlHelper sqlHelper, List<String> aliases, String sql) {
        this(sqlHelper, aliases, null, sql);
    }

    public SqlResourceType(ISqlHelper sqlHelper, List<String> aliases, List<ResourceType> child, String sql) {
        this.sqlHelper = sqlHelper;
        this.aliases = aliases;
        this.child = child;
        this.sql = sql;
    }

    @Override
    public String name() {
        return aliases.get(0);
    }

    @Override
    public List<String> aliases() {
        return aliases;
    }

    @Override
    public List<DbResource> read(Uri uri) throws BootException {

        final List<DbResource> objects = new ArrayList<>();
        final String sql = "select 'create table'";

        logger.debug(sql);

        final List<Map<String, String>> select = sqlHelper.select(sql);


        for (Map<String, String> row : select) {
            final List<String> objectsForUri = new ArrayList<>();
            final HashMap<String, String> objectHeaders = new LinkedHashMap<>();
            for (Map.Entry<String, String> column : row.entrySet()) {
                if (!column.getKey().startsWith("@")) {
                    objectsForUri.add(column.getValue());
                }
                objectHeaders.put(strip(column.getKey(), "@"), ofNullable(column.getValue()).orElse(""));
            }
            final String objectName = objectsForUri.get(objectsForUri.size() - 1);
            final DbResource object = new DbResourceThin(objectName, this, new DbUri(this.name(), objectsForUri),
                    objectHeaders);
            logger.debug("find object " + object.dbUri().toString());
            objects.add(object);
        }
        return objects;

    }

}

File: src/main/kotlin/com/github/mgramin/sqlboot/model/resourcetype/impl/sql/SqlResourceType.kt
/*
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2016-2019 Maksim Gramin
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.model.resourcetype.impl.sql

import com.github.mgramin.sqlboot.model.connection.DbConnection
import com.github.mgramin.sqlboot.model.resource.DbResource
import com.github.mgramin.sqlboot.model.resource.impl.DbResourceImpl
import com.github.mgramin.sqlboot.model.resourcetype.ResourceType
import com.github.mgramin.sqlboot.model.uri.Uri
import com.github.mgramin.sqlboot.model.uri.impl.DbUri
import com.github.mgramin.sqlboot.sql.select.SelectQuery
import com.github.mgramin.sqlboot.sql.select.impl.SimpleSelectQuery
import com.github.mgramin.sqlboot.sql.select.wrappers.JdbcSelectQuery
import com.github.mgramin.sqlboot.sql.select.wrappers.OrderedSelectQuery
import com.github.mgramin.sqlboot.sql.select.wrappers.PaginatedSelectQuery
import com.github.mgramin.sqlboot.template.generator.impl.GroovyTemplateGenerator
import org.apache.commons.lang3.StringUtils.strip
import reactor.core.publisher.Flux

/**
 * Created by MGramin on 12.07.2017.
 */
class SqlResourceType(
        private val aliases: List<String>,
        sql: String,
        private val connections: List<DbConnection>
) : ResourceType {

    private val simpleSelectQuery = SimpleSelectQuery(GroovyTemplateGenerator(sql))

    override fun aliases(): List<String> {
        return aliases
    }

    override fun path(): List<String> {
        return simpleSelectQuery.columns().keys
                .filter { v -> v.startsWith("@") }
                .map { v -> strip(v, "@") }
    }

    override fun read(uri: Uri): Flux<DbResource> {
        val mergeSequential: Flux<Map<String, Any>> =
                Flux.merge(
                        connections
                                .asSequence()
                                .map { connection ->
                                    return@map createQuery(uri, connection).execute(hashMapOf("uri" to uri))
                                            .map<Map<String, Any>?> {
                                                val mutableMap = it.toMutableMap()
                                                mutableMap["database"] = connection.getName()
                                                mutableMap
                                            }
                                }
                                .toList()

                )

        return mergeSequential
                .map { o ->
                    val path = o.entries
                            .filter { v -> v.key.startsWith("@") }
                            .map { it.value.toString() }
                    val name = path[path.size - 1]
                    val headers = o.entries
                            .map { strip(it.key, "@") to it.value }
                            .toMap()
                    DbResourceImpl(name, this, DbUri(this.name(), path), headers) as DbResource
                }
    }

    override fun metaData(): Map<String, String> {
        val columns = simpleSelectQuery.columns()
        val newColumns = columns.toMutableMap()
        newColumns["database"] = """{"label": "Database", "description": "Database name", "visible": true}"""
        return newColumns
    }

    private fun createQuery(uri: Uri, connection: DbConnection): SelectQuery {
        return JdbcSelectQuery(
                PaginatedSelectQuery(
                        OrderedSelectQuery(
                                simpleSelectQuery,
                                uri.orderedColumns()),
                        uri,
                        connection.paginationQueryTemplate()),
                dataSource = connection.getDataSource())
    }

}
